airgun/widgets.py

```diff
@@ -1267,6 +1287,96 @@ def read(self):
             "return ace.edit('{0}').getValue();".format(self.ace_edit_id))
 
 
+class Pagination(Widget):
+    """Represents Paginator widget that includes per page selector, First/Last/Next/Prev buttons
+    and current page index/overall amount of pages. Mainly used with Table widget.
+    """
+    ROOT = "//form[contains(@class, 'content-view-pf-pagination')]"
+    # Kattelo views use per_page with select, foreman use a per_page with Button DropDown.
+    PER_PAGE_BUTTON_DROPDOWN = ".//div[button[@id='pagination-row-dropdown']]"
+    PER_PAGE_SELECT = ".//select[contains(@ng-model, 'per_page')]"
+    first_page_button = Text(".//a[span[contains(@class, 'angle-double-left')]]")
+    previous_page_button = Text(".//a[span[contains(@class, 'angle-left')]]")
+    next_page_button = Text(".//a[span[contains(@class, 'angle-right')]]")
+    last_page_button = Text(".//a[span[contains(@class, 'angle-double-right')]]")
+    page = TextInput(locator=".//input[contains(@class, 'pagination-pf-page')]")
+    pages = Text(".//span[contains(@class, 'pagination-pf-pages')]")
+
+    def _element_exists(self, locator, parent=None, check_visibility=False):
+        """Check if a locator element/widget exists in the DOM"""
+        try:
+            exists = bool(self.browser.elements(
+                locator, parent=parent, check_visibility=check_visibility))
+        except NoSuchElementException:
+            exists = False
+        return exists
+
+    @property
+    def per_page(self):
+        """Return the per page widget"""
+        if self._element_exists(self.PER_PAGE_SELECT, parent=self):
+            return Select(self, self.PER_PAGE_SELECT)
+        return ButtonDropdown(self, self.PER_PAGE_BUTTON_DROPDOWN)
+
+    @property
+    def is_displayed(self):
+        """Check whether this Pagination widget exists and visible"""
+        return self._element_exists(self.pages, parent=self, check_visibility=True)
+
+    def _is_enabled(self, widget):
```

Comment:
This method is just bad. I know you just copied it, but let's rather not do that :)
- It's part of Paginator class, but actually checks other widgets, so it's more like helper which shouldn't be a part of paginator
- It checks for visibility of parent, not the widget itself, making its name basically misleading
- It doesn't work with widget, but uses pure selenium calls which we should try to avoid and use widgetastic methods where possible
- It messes with view/widget isolation, as it's looking outside of its scope

We can try pointing locators to `li` elements with that 'disabled' class, not the `a` inside of them. Then `self.browser.classes()` before click with no extra methods will be enough.

Reply:
completely removed